### PR TITLE
feat(crash): lazy Zarr loading for multi-GPU DDP training

### LINKED ISSUE #1550: 🚀[FEA]: Clarify DDP dataset memory behavior and recommend lazy loading for large Zarr datasets
### Is this a new feature, an improvement, or a change to existing functionality?

Improvement

### How would you describe the priority of this feature request

Medium

### Please provide a clear description of problem you would like to solve.

I want to report a DDP memory behavior that we validated and solved in practice.

Root cause:
- In multi-GPU DDP, DistributedSampler correctly shards sample indices across ranks.
- However, each rank/process still instantiates its own dataset object.
- In the crash data path, eager loading materializes large Zarr arrays during dataset construction.
- With 8 GPUs (8 ranks), this effectively multiplies host RAM usage by rank count and can lead to OOM/SIGKILL.

Key point:
- The problem is not DistributedSampler itself.
- The expensive part is eager dataset/reader initialization in CPU RAM.

Validated solution:
- Switching to lazy loading (on-demand reads during iteration) reduced CPU RAM significantly.
- After this change, distributed training became stable and scalable.

Request for PhysicsNeMo:
1. Explicitly document in the crash/DDP example that DistributedSampler shards indices only, not dataset memory.
2. Recommend lazy loading for large Zarr datasets in multi-GPU runs.
3. Optionally add guidance for rank-aware early sharding of files/runs to reduce duplicate work across ranks.

Relevant paths in repo:
- examples/structural_mechanics/crash/train.py
- examples/structural_mechanics/crash/datapipe.py
- examples/structural_mechanics/crash/zarr_reader.py

### Describe any alternatives you have considered

- Lowering nproc_per_node: reduces memory pressure but also reduces throughput.
- Setting num_workers=0: helps worker overhead but does not remove rank-level dataset duplication.
- Keeping eager loading: still causes high RAM in multi-GPU DDP.
- Lazy loading: this is the solution that worked in our tests.

### PR DESCRIPTION
**Pull request by Thabhelo**
## Summary
- Add lazy Zarr loading (`lazy_load: true` by default) so mesh trajectories and point features materialize on first sample access instead of at dataset construction.
- Update the crash datapipe to stream stats and defer per-sample tensors/graphs when lazy records are returned.
- Document DDP memory behavior (`DistributedSampler` shards indices, not host RAM) in the crash README and log a hint in `train.py` when multi-GPU + lazy Zarr are active.

closes #1550

## Test plan
- [x] `pytest examples/structural_mechanics/crash/tests/test_zarr_reader.py examples/structural_mechanics/crash/tests/test_datapipe_lazy.py`
- [x] Eager-mode smoke test with `Reader(lazy_load=False)` (all samples materialized at init)
- [x] pre-commit on changed files

## Notes
- This PR is scoped to #1550 (lazy Zarr + DDP docs). A few short docstrings on existing crash example symbols were added only as a **temporary pre-commit workaround** — the interrogate hook falsely flagged baseline-grandfathered items due to a path-parsing bug in `check_docstring_coverage.py` (details in [follow-up comment](https://github.com/NVIDIA/physicsnemo/pull/1703#issuecomment-4637334934)).
- Root-cause fix tracked in #1704.

### CODE REVIEW

**Comment by Thabhelo** [comment]
### Follow-up: interrogate hook false positives (separate PR)

**What this PR is for:** lazy Zarr loading and DDP memory documentation for #1550.

While getting this through pre-commit, the **interrogate** check failed even though the flagged symbols (`SimSample.to`, `CrashGraphDataset.create_graph`, `Trainer.train`, etc.) were already listed in `test/ci_tests/interrogate_baseline.txt`.

**Root cause:** `test/ci_tests/check_docstring_coverage.py` parses interrogate section headers with a regex that only matches lines surrounded by **4+ equals signs** (`={4,}`). Interrogate 1.7 (pinned in pre-commit) prints headers like:

```
= Coverage for /abs/path/examples/structural_mechanics/crash/ =
```

Because that header is not matched, `current_dir` stays empty and file paths are resolved incorrectly (e.g. `../../../datapipe.py` instead of `examples/structural_mechanics/crash/datapipe.py`). Baseline entries then look like "new" undocumented items whenever those example files are edited.

**Temporary workaround in this PR:** added small docstrings on those symbols so they are no longer reported as `MISSED` and the hook passes.

**Actual fix:** #1704 broadens the header parser so baseline matching works correctly — this means no docstring workarounds needed going forward even for the next person who touches an example file with baseline gaps.

**Comment by Thabhelo** [comment]
Follow-up parser fix opened in #1704.

**Comment by Thabhelo** [comment]
Reviewed PR comments: triage:

- **Greptile P1 (lazy feature/edge running mean):** Valid. Fixed in 0b5e96a — lazy stat helpers now use the same `zeros` + `m/num_samples` accumulation as the eager path (first sample was previously overweighted).
- **Greptile P1 (edge stats on raw vs normalized positions):** Valid for graph + lazy path. Fixed in 0b5e96a — `_compute_edge_stats_lazy` now materializes normalized positions before building temporary graphs, matching eager construction order.
- **Greptile P2 (eager double-normalize landmine):** Valid defensive fix. Addressed in 0b5e96a — non-lazy `_load_sample_tensors` returns already-normalized tensors without re-applying normalization.

Thanks @greptile-apps for the catch.

**Comment by Thabhelo** [comment]
Good catch, fixed in 349312a. `_compute_edge_stats` now uses the same `zeros + m/num_samples` accumulation as `_compute_edge_stats_lazy`.

**Comment by Thabhelo** [comment]
> Thanks for the detailed triage and for addressing all three issues. The lazy stat accumulation, normalized-position edge stats, and eager double-normalize fixes in `0b5e96a` all look correct.
> 
> One remaining issue worth flagging: the **eager `_compute_edge_stats`** still has the original first-sample overweighting bug that was fixed in the lazy path. The sentinel `None` pattern leaves `m_0` with full weight while `m_1…m_{N-1}` are each divided by `N`:
> 
> ```python
> # datapipe.py – _compute_edge_stats (eager)
> edge_mean = m if edge_mean is None else edge_mean + m / self.num_samples
> #           ↑ full weight for i=0; all others get 1/N weight
> ```
> 
> This should use the same `zeros + m/N` pattern as `_compute_edge_stats_lazy`:
> 
> ```python
> def _compute_edge_stats(self):
>     edge_mean = None
>     edge_meansqr = None
>     edge_dim = None
>     for i in range(self.num_samples):
>         x_e = self.graphs[i].edge_attr.to(torch.float32)  # [E,De]
>         if edge_dim is None:
>             edge_dim = x_e.shape[1]
>             edge_mean = torch.zeros(edge_dim, dtype=torch.float32)
>             edge_meansqr = torch.zeros(edge_dim, dtype=torch.float32)
>         m = torch.mean(x_e, dim=0)
>         msq = torch.mean(x_e * x_e, dim=0)
>         edge_mean += m / self.num_samples
>         edge_meansqr += msq / self.num_samples
> ```
> 
> The same sentinel pattern also appears in `_compute_feature_stats`, but that one is fine — `feat_mean` is initialized as `zeros(fdim)` before the loop and accumulates uniformly. Only `_compute_edge_stats` has the stale `None`-sentinel issue.
> 
> Tip: You can customize Greptile's behavior for this repo with [`.greptile/rules.md` and `.greptile/config.json`](https://www.greptile.com/docs/code-review/custom-standards).

This is a pre-existing bug (not introduced by lazy loading), but you are right @greptile-apps that it’s still in the file we’re editing.

File: examples/structural_mechanics/crash/datapipe.py [hunk 1]
```diff
@@ -415,6 +362,213 @@ def build_xy(self, batch_idx: int, time_idx: int | None):
...
+            _, feats, _ = self._load_sample_tensors(i, retain=False, normalize=False)
+            x = feats.to(torch.float32)
+            if fdim is None:
+                fdim = x.shape[1]
+            assert x.shape[1] == fdim, f"Feature dim mismatch: {x.shape[1]} vs {fdim}"
+            m = torch.mean(x, dim=0)
+            msq = torch.mean(x * x, dim=0)
+            feat_mean = m if feat_mean is None else feat_mean + m / self.num_samples
+            feat_meansqr = (
+                msq if feat_meansqr is None else feat_meansqr + msq / self.num_samples
```
[the remark this answers is not in the document]
Comment (Thabhelo) [reply examples/structural_mechanics/crash/datapipe.py]: Fixed in 0b5e96a. Lazy stat helpers now use zeros plus m/num_samples, same as the eager path.

File: examples/structural_mechanics/crash/datapipe.py [hunk 2]
```diff
@@ -415,6 +362,213 @@ def build_xy(self, batch_idx: int, time_idx: int | None):
...
+    def _load_sample_tensors(
+        self, batch_idx: int, *, retain: bool = True, normalize: bool = True
+    ) -> tuple[torch.Tensor, torch.Tensor, dict[str, torch.Tensor]]:
+        if not self._lazy_mode:
+            mesh = self.mesh_pos_seq[batch_idx]
+            feats = self.node_features_data[batch_idx]
+            targets = self.target_series_data[batch_idx]
+            if normalize:
+                mesh, feats = self._normalize_sample_tensors(mesh, feats)
+            return mesh, feats, targets
```
[the remark this answers is not in the document]
Comment (Thabhelo) [reply examples/structural_mechanics/crash/datapipe.py]: Fixed in 0b5e96a. Eager samples skip re-normalization in _load_sample_tensors.

**Comment by Thabhelo** [comment]
@coreyjadams gentle ping when you have a moment. Greptile items are addressed and replied on the threads. Closes #1550. Thanks!
